sky/skylet/subprocess_daemon.py

```diff
@@ -1,17 +1,35 @@
 """Sky subprocess daemon.
-
 Wait for parent_pid to exit, then SIGTERM (or SIGKILL if needed) the child
 processes of proc_pid.
 """
-
 import argparse
+import os
 import sys
 import time
 
 import psutil
 
-if __name__ == '__main__':
 
+def daemonize():
+    """Separates the process from its parent process with double-forking."""
+    # First fork
+    if os.fork() > 0:
+        # original process terminates.
+        sys.exit()
+
+    # Continues to run from first forked child process.
+    # Detach from parent environment
+    os.setsid()
+
+    # Second fork
+    if os.fork() > 0:
+        # The first forked child process terminates.
+        sys.exit()
+    # Continues to run from second forked child process.
```

Comment:
It seems we have already set `start_new_session=True` when we start this daemon, which should already do `setsid` for this process. Do we know why that does not work?
https://github.com/skypilot-org/skypilot/blob/2e545b8db70991bc74d112d5070fa908078f9cdb/sky/skylet/log_lib.py#L211-L219

Reference: https://docs.python.org/3/library/subprocess.html
> If start_new_session is true the setsid() system call will be made in the child process prior to the execution of the subprocess.

Reply:
So the `start_new_session=True` flag certainly does start a new session by running `setsid`. See the three comparison below:


1. Running a `ray` job with the flag, `start_new_session=True`, and without the daemonization for `subprocess_daemon.py` I added(what we currently have at `master` branch):
```
$ ps aux | grep subprocess_daemon
gcpuser     4536  0.0  0.0  16212 12704 ?        SNs  18:37   0:00 /home/gcpuser/skypilot-runtime/bin/python /home/gcpuser/skypilot-runtime/lib/python3.10/site-packages/sky/skylet/subprocess_daemon.py --parent-pid 4141 --proc-pid 4532
gcpuser     4885  0.0  0.0   5264   636 pts/0    S+   18:39   0:00 grep subprocess_daemon
$ ps -o pid,ppid,sid,cmd 4536
    PID    PPID     SID CMD
   4536    4141    4536 /home/gcpuser/skypilot-runtime/bin/python /home/gcpuser/skypilot-runtime/lib/python3.10/site-pa
$ ps -o pid,ppid,sid,cmd 4141
    PID    PPID     SID CMD
   4141    3710    3527 ray::task,
```
As you can see from above, the `SID`, session id, is different for the process running `subprocess_daemon.py` and `ray::task`(`4536` and `3527`). And this is due to the `start_new_session=True` flag. And you can see that process running `subprocess_daemon.py` is a child process of the `ray::task` process, which kills the daemon process when `ray`'s `stop_job` API is ran.


2. Running a `ray` job with the flag, `start_new_session=True`, and daemonization for `subprocess_daemon.py` I added(what we currnetly have at this PR):

```
$ ps aux | grep subprocess_daemon
gcpuser     4531  0.3  0.0  16228  9872 ?        SN   18:08   0:00 /home/gcpuser/skypilot-runtime/bin/python /home/gcpuser/skypilot-runtime/lib/python3.10/site-packages/sky/skylet/subprocess_daemon.py --parent-pid 4129 --proc-pid 4522
gcpuser     5029  0.0  0.0   5132   708 pts/0    S+   18:11   0:00 grep subprocess_daemon
$ ps -o pid,ppid,sid,cmd 4531
    PID    PPID     SID CMD
   4531       1    4526 /home/gcpuser/skypilot-runtime/bin/python /home/gcpuser/skypilot-runtime/lib/python3.10/site-pa
$ ps -o pid,ppid,sid,cmd 4129
    PID    PPID     SID CMD
   4129    3699    3516 ray::task,
```
As you can see from above, the `SID` is different(`4526` and `3516`) for the process running `subprocess_daemon.py` and `ray::task`. Also, the parent process id, `PPID`, of the process running `subprocess_daemon.py` is not `ray::task` anymore with the daemonization added from this PR. As the `subdaemon_process.py` process is completely detached from it's original parent process, `ray::task`, the `subprocess_daemon.py` process does not get killed anymore when `ray`'s `stop_job` API is ran.


3. Running a `ray` job without the flag, `start_new_session=True`, but with daemonization for `subprocess_daemon.py` I added(just to show session group is separated with `start_new_session=True` flag):
```
$ ps aux | grep subprocess_daemon
gcpuser     4553  0.2  0.0  16228  9876 ?        SN   18:18   0:00 /home/gcpuser/skypilot-runtime/bin/python /home/gcpuser/skypilot-runtime/lib/python3.10/site-packages/sky/skylet/subprocess_daemon.py --parent-pid 4147 --proc-pid 4544
gcpuser     4650  0.0  0.0   5132   640 pts/0    S+   18:18   0:00 grep subprocess_daemon
$ ps -o pid,ppid,sid,cmd 4553
    PID    PPID     SID CMD
   4553       1    3502 /home/gcpuser/skypilot-runtime/bin/python /home/gcpuser/skypilot-runtime/lib/python3.10/site-
$ ps -o pid,ppid,sid,cmd 4147
    PID    PPID     SID CMD
   4147    3685    3502 ray::task,
```
As you can see from above, the `SID` is identical(`3502`) for both `ray::task` and `subprocess_deamon.py` as I removed the `start_new_session=True` flag, so the flag is indeed creating a new session when it's launching `subprocess_daemon.py` as a process.


In my understanding, creating a process with a separate session group doesn't necessarily detaches the process(`subprocess_daemon.py`) from it's parent process(`ray::task`) in terms of process hierarchy. So the additional daemonization is necessary.

Reply:
Thanks for the comprehensive investigation for this @landscapepainter ! This is very useful! I am wondering if it's possible that we can add some additional arguments to the subprocess call in python to achieve this detachment. That might be cleaner than having the current handling in the `subprocess_daemon.py` because the fork can cause two processes (though one will be exited immediately) to be generated every time we start this demon process?

Reply:
@Michaelvll I did some extensive search on this before I implemented for current solution, and just looked into more options now in case I missed any.

The observed reason why `ray`'s `stop_job` api kills `subprocess_daemon.py` is because the process is a child process of `ray::task` within the process hierarchy although the session group and process group are different, which implies the possible solution is to run `subprocess_daemon.py` in a separate process hierarchy from `ray::task` process. Unfortunately, there isn't an option for `subprocess.Popen()` to completely detach the child process from the process hierarchy.

Another approach that seems feasible is to update the way how we create ray job scripts from RayCodeGen and submit jobs to perhaps run the subprocess_daemon.py separately from ray job. But this seems to include much larger refactoring task since now skypilot needs access to the process IDs generated within `ray::task` to pass as an argument to and run `subprocess_daemon.py` outside of the job script.

Reply:
That makes sense. If we `daemonize` it in this file already, do we still need to add `start_new_session` in our `Popen` call?

Also, please leave a comment at the top of this file or this function to say why we daemonize this process, e.g., to make sure `ray job stop` does not kill this process, so this process can handle any leaked processes from user program.

Reply:
`start_new_session` is redundant at this point as you mentioned, and confirmed it works fine without it. Removed the option, and also updated the docstring focusing more on why.